File: lib/services/streak_manager.dart
import 'storage_service.dart';

/// Manages daily streak tracking for the application
class StreakManager {
  static final StreakManager _instance = StreakManager._internal();
  factory StreakManager() => _instance;
  StreakManager._internal();

  final StorageService _storage = StorageService();

  static const String _lastCompletionDateKey = 'last_completion_date';
  static const String _currentStreakKey = 'current_streak';
  static const String _longestStreakKey = 'longest_streak';

  /// Record a task completion and update streak
  Future<int> recordTaskCompletion() async {
    final today = _getTodayDateString();
    final lastCompletion = await _storage.getString(_lastCompletionDateKey);
    final currentStreak = await getCurrentStreak();

    int newStreak = currentStreak;

    if (lastCompletion == null) {
      // First time completion
      newStreak = 1;
    } else if (lastCompletion == today) {
      // Already completed today, no change
      return currentStreak;
    } else {
      final lastDate = DateTime.parse(lastCompletion);
      final todayDate = DateTime.parse(today);
      final difference = todayDate.difference(lastDate).inDays;

      if (difference == 1) {
        // Consecutive day
        newStreak = currentStreak + 1;
      } else {
        // Streak broken, reset to 1
        newStreak = 1;
      }
    }

    // Save the new streak
    await _storage.saveInt(_currentStreakKey, newStreak);
    await _storage.saveString(_lastCompletionDateKey, today);

    // Update longest streak if applicable
    final longestStreak = await getLongestStreak();
    if (newStreak > longestStreak) {
      await _storage.saveInt(_longestStreakKey, newStreak);
    }

    return newStreak;
  }

  /// Get the current streak count
  Future<int> getCurrentStreak() async {
    final lastCompletion = await _storage.getString(_lastCompletionDateKey);

    if (lastCompletion == null) {
      return 0;
    }

    final today = _getTodayDateString();
    final lastDate = DateTime.parse(lastCompletion);
    final todayDate = DateTime.parse(today);
    final difference = todayDate.difference(lastDate).inDays;

    // If last completion was more than 1 day ago, streak is broken
    if (difference > 1) {
      await resetStreak();
      return 0;
    }

    // Otherwise return the stored streak
    final streak = await _storage.getInt(_currentStreakKey);
    return streak ?? 0;
  }

  /// Get the longest streak ever achieved
  Future<int> getLongestStreak() async {
    final streak = await _storage.getInt(_longestStreakKey);
    return streak ?? 0;
  }

  /// Reset the streak to 0
  Future<void> resetStreak() async {
    await _storage.saveInt(_currentStreakKey, 0);
    await _storage.remove(_lastCompletionDateKey);
  }

  /// Check if user has completed a task today
  Future<bool> hasCompletedToday() async {
    final lastCompletion = await _storage.getString(_lastCompletionDateKey);
    if (lastCompletion == null) return false;

    final today = _getTodayDateString();
    return lastCompletion == today;
  }

  /// Get the last completion date
  Future<DateTime?> getLastCompletionDate() async {
    final lastCompletion = await _storage.getString(_lastCompletionDateKey);
    if (lastCompletion == null) return null;

    try {
      return DateTime.parse(lastCompletion);
    } catch (e) {
      print('Error parsing last completion date: $e');
      return null;
    }
  }

  /// Get streak statistics
  Future<Map<String, dynamic>> getStreakStats() async {
    final currentStreak = await getCurrentStreak();
    final longestStreak = await getLongestStreak();
    final hasCompletedToday = await this.hasCompletedToday();
    final lastCompletion = await getLastCompletionDate();

    return {
      'currentStreak': currentStreak,
      'longestStreak': longestStreak,
      'hasCompletedToday': hasCompletedToday,
      'lastCompletionDate': lastCompletion,
    };
  }

  /// Check if streak milestone reached (e.g., 7, 30, 100 days)
  bool isStreakMilestone(int streak) {
    const milestones = [7, 14, 30, 50, 100, 200, 365];
    return milestones.contains(streak);
  }

  /// Get today's date as string in YYYY-MM-DD format
  String _getTodayDateString() {
    final now = DateTime.now();
    return '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}';
  }
}

File: eredeti_figma/src/utils/streakManager.ts
interface StreakData {
  lastCompletionDate: string; // YYYY-MM-DD format
  currentStreak: number;
  completedToday: boolean;
}

const STREAK_STORAGE_KEY = 'daily_streak';

// Get today's date in YYYY-MM-DD format
function getTodayString(): string {
  const today = new Date();
  return today.toISOString().split('T')[0];
}

// Get yesterday's date in YYYY-MM-DD format
function getYesterdayString(): string {
  const yesterday = new Date();
  yesterday.setDate(yesterday.getDate() - 1);
  return yesterday.toISOString().split('T')[0];
}

// Save streak data to localStorage
function saveStreakData(data: StreakData): void {
  try {
    localStorage.setItem(STREAK_STORAGE_KEY, JSON.stringify(data));
  } catch (error) {
    console.error('Error saving streak data:', error);
  }
}

// Load streak data from localStorage
export function loadStreakData(): StreakData {
  try {
    const saved = localStorage.getItem(STREAK_STORAGE_KEY);
    if (saved) {
      const data = JSON.parse(saved) as StreakData;
      const today = getTodayString();
      const yesterday = getYesterdayString();
      
      // Check if the saved date is today - if yes, keep completedToday flag
      // If not today, reset completedToday to false for the new day
      if (data.lastCompletionDate !== today) {
        data.completedToday = false;
        
        // If last completion was NOT yesterday, streak is broken
        if (data.lastCompletionDate !== yesterday) {
          console.log('📅 Streak expired. Last completion:', data.lastCompletionDate);
          data.currentStreak = 0;
          data.lastCompletionDate = '';
          // Save the reset data
          saveStreakData(data);
        }
      }
      
      return data;
    }
  } catch (error) {
    console.error('Error loading streak data:', error);
  }
  
  return {
    lastCompletionDate: '',
    currentStreak: 0,
    completedToday: false
  };
}

// Record a task completion and update streak
export function recordTaskCompletion(): { isFirstToday: boolean; newStreak: number } {
  const data = loadStreakData();
  const today = getTodayString();
  const yesterday = getYesterdayString();
  
  console.log('🔥 recordTaskCompletion:', {
    today,
    yesterday,
    lastCompletionDate: data.lastCompletionDate,
    currentStreak: data.currentStreak,
    completedToday: data.completedToday
  });
  
  // If already completed today, just return current streak
  if (data.completedToday) {
    console.log('✅ Already completed today, returning current streak:', data.currentStreak);
    return {
      isFirstToday: false,
      newStreak: data.currentStreak
    };
  }
  
  // This is the first completion today
  let newStreak = 1;
  
  if (data.lastCompletionDate === yesterday) {
    // Completed yesterday, so increment streak
    newStreak = data.currentStreak + 1;
    console.log('📈 Completed yesterday! Incrementing streak:', newStreak);
  } else if (data.lastCompletionDate === today) {
    // Already completed today (shouldn't happen, but just in case)
    newStreak = data.currentStreak;
    console.log('✅ Already completed today (edge case)');
  } else {
    // Streak broken or first time, reset to 1
    newStreak = 1;
    console.log('🔄 Streak broken or first time. Starting new streak:', newStreak);
  }
  
  // Update and save
  const updatedData: StreakData = {
    lastCompletionDate: today,
    currentStreak: newStreak,
    completedToday: true
  };
  
  saveStreakData(updatedData);
  console.log('💾 Saved new streak data:', updatedData);
  
  return {
    isFirstToday: true,
    newStreak: newStreak
  };
}

// Get current streak count
export function getCurrentStreak(): number {
  const data = loadStreakData();
  const today = getTodayString();
  const yesterday = getYesterdayString();
  
  console.log('📊 getCurrentStreak check:', {
    today,
    yesterday,
    lastCompletionDate: data.lastCompletionDate,
    currentStreak: data.currentStreak,
    completedToday: data.completedToday
  });
  
  // If last completion was today or yesterday, return current streak
  if (data.lastCompletionDate === today || data.lastCompletionDate === yesterday) {
    return data.currentStreak;
  }
  
  // Streak is broken - reset streak data
  if (data.currentStreak > 0) {
    console.warn('⚠️ Streak broken! Resetting to 0');
    const resetData: StreakData = {
      lastCompletionDate: '',
      currentStreak: 0,
      completedToday: false
    };
    saveStreakData(resetData);
  }
  
  return 0;
}

// Reset streak (for game reset/restart)
export function resetStreak(): void {
  try {
    localStorage.removeItem(STREAK_STORAGE_KEY);
  } catch (error) {
    console.error('Error resetting streak data:', error);
  }
}
